### PR TITLE
Add SotA to Mainline

### PR DESCRIPTION
**Pull request by sigurdfdragon**
Adds the campaign 'Secrets of the Ancients' by beetlenaut (Dan Gerhards) to mainline.

Not yet ready for merging.

### CODE REVIEW

**Comment by Wedge009** [comment]
Oh new campaign, nice.

What is the campaign rank supposed to do?

**Comment by gfgtdf** [comment]
was there any previous discussion on this? Specially with the orignal author and with members of the dev team?

What is the size of this campaign on disk?

**Comment by CelticMinstrel** [comment]
> What is the size of this campaign on disk?

...why does this matter?

> Is there a plan to remvoe on o the other campagsn when we add this one?

Why would we remove a campaign? There's no particular limit to the number of campaigns mainline is allowed to have.

> What is the campaign rank supposed to do?

It pushes the campaign up towards the top of the list.

**Comment by sigurdfdragon** [comment]
@Wedge009 - rank determines what order the campaign is shown in the campaign list. As it is an intermediate difficulty campaign, I added it at the end of that section.

@gfgtdf - Yes, with both author and dev team, though the discussion was mostly during the summer.
My size on disk in Win 7 is 7.44 MB
There isn't a github repo of the campaign, I did ask beetlenaut about it.
No plans to remove a campaign as far as I'm aware.

**Comment by CelticMinstrel** [comment]
There are a few images that might be better merged into core (though the only one that jumped out at me was the open chest, which if I recall correctly has a closed analogue already in core).

I wonder if it's really impossible to make the ship work without customizing the bats? If this is really the case though, would it be okay to apply those customization to the core flying units rather than using customized versions of them in this campaign? This is a question because I'm not sure if that might have other negative side-effects.

**Comment by gfgtdf** [comment]
Hmm i don't have objections on this, adding but i don't really know this campaign eigher.

It seems like this is indendet with tabs, while mainline wml content usually gets indented with spaces. please run wmlindent over it.

**Comment by beetlenaut** [comment]
All flying units *used* to be displayed behind terrain that stuck up from a 
tile below them on the screen like great trees, for example. That was 
changed--I think for 1.12. Now they are always drawn in front of terrain no 
matter what the z-level of the terrain is. This looks strange in the case of 
great trees IMO, but it looked a lot worse for the ship. Since all the flying 
units were changed at the same time, it was obviously considered a desirable 
feature, so I just worked around it with my own bats in SotA. There must have 
been some discussion about this change before it happened, but I didn't see 
it.

---Dan

**Comment by ln-zookeeper** [comment]
From looking around for a bit, there seems to be all sorts of little issues. Maybe some of them are intentional and aren't a problem, but some are pretty immediately apparent as being problems (even if not huge ones).

- Bigmap background missing
- Weird void in the castle in S1
- Recruiting corpses in S2 via a custom menu item instead of the recruit menu
- The cemetery fence is built of two huge images; no idea if that's a problem
- The "You can now recruit skeletons" message image seems bugged
- There is no "Right-click to simplify the ship graphics" menu item
- Unused or unnecessary images (items/cliff-wall, items/barrel, two versions of items/gold-coins)
- Map files would be good to be numbered
- Why are there custom frames and a unit type for the sacrifice animation?
- Why does turning Necromancers to Liches happen through a custom menu item instead of being an advancement option?
- Isthiniel only has a couple of lines but has a weird custom portrait
- The screenshot-based story images don't age well, in fact they're already mismatched WRT water
- Objectives ending with a period and maybe a few other formatting issues

**Comment by beetlenaut** [comment]
On Wednesday, December 07, 2016 12:52:55 AM Lari Nieminen wrote:
> - Bigmap background missing
It shouldn't be. It's just titlescreen.png. Is the path wrong in sota-
utils.cfg?

> - Weird void in the castle in S1
I didn't want to use keeps for the inside of the school because you're not 
supposed to be able to recruit from there, or reach it at all in fact. I just 
didn't have a better idea than making it black. If someone does have a better 
idea, we can definitely use it. It might be best with a custom image to 
represent the roof and walls of the school building, but that's beyond me to 
draw.

> - Recruiting corpses in S2 via a custom menu item instead of the recruit
It doesn't work like normal recruiting. There is no castle--you have to use a 
specific, unused grave instead. To me, the custom menu item seems like the 
natural choice here.

> menu - The cemetery fence is built of two huge images; no idea if that's
> a problem
I don't see why it would be. But, if a terrain maestro wants to turn 
graveyard_fence and graveyard_gate into normal terrains, lots of people would 
appreciate it I'm sure! They are at different z-levels though.

 > - The "You can now recruit skeletons" message image seems bugged
Yeah, the images changed size, and that one slipped past me. FIXED on the add-
on server. You can get the correct line from there.

> - There is no "Right-click to simplify the ship graphics" menu item
I don't know where this went! I'm pretty sure it used to work, and the same 
code still adds the menu item just fine on 1.12. Did something change in the 
menu item system for 1.13? The wiki doesn't say anything about it. I need help 
with this one.

> - Unused or unnecessary images (items/cliff-wall, items/barrel, two
> versions of items/gold-coins)
The barrel *is* used, and I think it's better than the core one. The two coins 
are both used too--one is a pile of coins on the floor in a map, and the other 
is used in a dialog box. As such, it has a spacer on the bottom. This could 
probably be handled better by blitting the other image onto a transparent 
background, but that takes a lot of trial and error. I did find three 
unnecessary files though: maps/temp.map, items/cliff-wall.png, and 
terrain/cliff-wall.png. units/zombie-wolf.png is not used either, but it could 
be a starting point for someone wanting to make that variation.

> - Map files would be good to be numbered
Is this really a problem? Dead Water doesn't have numbered maps either, and 
nobody ever said anything about it. It wouldn't take that long to do, it would 
just be tedious.

> - Why are there custom frames and a unit type for the sacrifice animation?
The blade swing has to stop before where it does during the attack of the 
normal unit. Using a different unit type seemed easy enough. Was there a 
better way to do it?

> - Why does turning Necromancers to Liches happen through a custom menu
> item instead of being an advancement option?
It *is* an advancement option after this point in the campaign, but it wasn't 
before. If you have a level-3 necromancer, you already missed the option to 
choose the lich path for its advancement. The menu item lets you change your 
mind about the advancement path if you want to, now that it's been unlocked.

> - Isthiniel only has a couple of
> lines but has a weird custom portrait
It's from the card game. I was happy to use something most people haven't seen 
before, but if it's too weird, we can use a normal elf portrait instead.

> - The screenshot-based story images don't
> age well, in fact they're already mismatched WRT water
Sure. They are just the best I could do with my limited art skills. I think it 
would be better to leave them in for now until we can get an artist to do 
custom story screens. They are probably at least a little better than nothing 
at all.

> - Objectives ending with a period
That was me being an English teacher. If it breaks the style, I guess they 
should go.

> maybe a few other formatting issues
If so, it was not intentional. I'm sorry if this makes extra work for someone.

Something that bugs *me* is the lack of wolf and rat zombie variations. I was 
never successful in drawing either of them. I also wasn't able to do a very 
good job of animating the custom Ardryn-na sprite.

---Dan

**Comment by Vultraz** [comment]
> It doesn't work like normal recruiting. There is no castle--you have to use a specific, unused grave instead.
There is a keep/castle overlay terrain you can use to enable recruiting on non-castle terrain.

> The barrel *is* used, and I think it's better than the core one. 
The barrel image is the same one that's in mainline in 1.13.

> Dead Water doesn't have numbered maps either, and nobody ever said anything about it.
Likely because it seems to use the same map for multiple scenarios. Otherwise, numbered map files is preferred.

**Comment by beetlenaut** [comment]
> There is a keep/castle overlay terrain you can use to enable recruiting on
> non-castle terrain.
I know that, but the graveyard doesn't work like a castle. You can only use 
each grave once, and your leader has to be adjacent to that specific grave to 
use it. Using castle terrains would require turning them on and off 
continuously for moves, undo/redo, recruits and so on. And most of the time, 
you can only get one kind of unit anyway. I don't see any advantage. In fact, 
I think that using an invisible castle might be more confusing for the player 
rather than less.

> The barrel image is the same one that's in mainline in 1.13.
Oh! I didn't know it got committed. Nobody mentioned it. So, that image is not 
actually necessary then.

> numbered map files is preferred.
Okay.

---Dan

**Comment by CelticMinstrel** [comment]
Regarding unnecessary images, I'd say those should never be committed, rather than being committed and then deleted in a later commit. In other words, rewrite the initial commit to omit them.

Does anyone else agree with me about promoting the open chest to core?

**Comment by Vultraz** [comment]
I see no reason it cannot be in core.

**Comment by ln-zookeeper** [comment]
Yeah, the open chest is fine. I've added it to core now.

**Comment by beetlenaut** [comment]
On Wednesday, December 07, 2016 12:52:55 AM Lari Nieminen wrote:
> - There is no "Right-click to simplify the ship graphics" menu item

I found the problem. Apparently, "equals=true" doesn't work any more. You have 
to use "boolean_equals=true". I fixed sota-utils so the menu item appears, and 
also fixed a small bug in the next scenario, Blackwater, where the sails 
didn't start out hidden if they had been hidden in the previous scenario.

So there is a new sota-utils.cfg and 05_Blackwater.cfg on the add-ons server.

---Dan

**Comment by CelticMinstrel** [comment]
`equals=true` tests for string equality, so if the value it's comparing to happens to be "yes" rather than "true", it'll fail.

**Comment by beetlenaut** [comment]
I understand that, but I only used "true" and "false".  And like I said, the 
same code worked in 1.12 (comparing as strings), so this is new behavior. If 
it was not intentional, it must be a bug with the engine.

---Dan

**Comment by gfgtdf** [comment]
> so this is new behavior

Whe wml tables gets converted to c++ structs or to lua tables the 'true' string will automatically be converted to the c++/lua bool value true, which then when converting back to a wml table will be writen as 'yes'. In particular when a tag implementation gets converted to lua, like it happend with multiple tags during 1.13.6, this means that whenever you use 'true' in those tags it migth happen that it s silently gets converted to 'yes'

**Comment by sigurdfdragon** [comment]
Unfortunately, I no longer have sufficient time to see this project through properly. As such, I cannot continue to lead mainlining efforts for this campaign. I apologize for any inconvenience this has caused.

**Comment by CelticMinstrel** [comment]
Should we close this PR then, or wait until someone else opens a new PR for it?

**Comment by beetlenaut** [comment]
Then it should definitely be noted in the wiki that you should always use 
"boolean_" with boolean comparisons, or that you should always use "yes" or 
"no", or whatever you want people to do now. This took me a long time to 
figure out with just trial and error, and I am probably not the only one who 
has been using string comparisons for a long time with no trouble.

---Dan

**Comment by beetlenaut** [comment]
I have some free time myself, so I could take over and fix the various 
problems that have cropped up. I haven't used git or github before though, and 
reading about it has been surprisingly unhelpful so far. Should I just start 
poking at things and hope I don't break them, or does someone want to give me 
some tips? At the moment I don't know how I would make changes to the PR or 
even download it in its current state.

---Dan

**Comment by Pentarctagon** [comment]
To download the PR, you'd use `git clone https://github.com/sigurdfdragon/wesnoth-fork.git <path/to/where/you/want/fork/downloaded> --branch "(WIP)Add_SotA_to_Mainline"`.

To make changes to the PR, you would:
1) Make your change(s) locally in the branch you downloaded
2) View the list of altered/added/removed files with `git status`, or you can show the actual changes with `git diff`
3) Add any files you want to commit with `git add <list> <of> <files> <or> <folders>`
4) Commit your changes locally with `git commit`
4a) Instead of using `git add`, you can use `git commit -a`, however this won't commit new files
5) Add the commit message
5a) This can be done via `git commit -m "short summary" [-m "long description"]`
5b) This can also be done by omitting the `-m` options, which should then open the configured text editor (changed via `git config --global core.editor <text editor>`) to let you enter your commit message that way.  This will contain a commented list of your changes, and you enter your summary and description below.  For example:
> \# s
> \# t
> \# u
> \# f
> \# f
> Summary
> 
> Description
6) Push the commit with `git push`
6a) This will error the first time with an error asking who you are.  You will need to provide `git config --global user.email "you@example.com"` and `git config --global user.name "Your Name"`

There is also some example/advice on Wesnoth's [wiki](https://wiki.wesnoth.org/WesnothRepository).

**Comment by Pentarctagon** [comment]
Also, probably not an issue, but assuming that relicensing mainline resources as CC-BY-SA ever starts moving forward again, would that cause any issues with getting any of the images relicensed?

Then again, if that effort has completely petered out, then I guess it doesn't matter at all.  There haven't actually been any updates on that since the end of October, and even the more recent notes from the Board IRC didn't show any particular progress being made on a decision.

**Comment by Wedge009** [comment]
@beetlenaut I recall there were some changes to Lua in the last few months and part of those changes were reflected in the yes/true behaviour that you noticed. I think @GregoryLundberg already resolved the affected instances in the main campaigns - obviously the individual add-on maintainers would need to check for this in their own campaigns.

Anyway, looking forward to seeing this in the next release.

**Comment by CelticMinstrel** [comment]
Pentarctagon's instructions are quite thorough, but I'll add that most of what he described can also be done with `git gui`, run from within the wesnoth source directory. You can also reduce download size when cloning by adding something like `--depth=50`.

Interesting that he says you put the commit message below the summary. I've always put it above. I'm pretty sure it wouldn't make any difference, though.

**Comment by beetlenaut** [comment]
I don't have strong feelings about the license or use of my images, so it's OK 
if the license gets changed. However, the majority of the images in this 
campaign are modifications of other people's images, and I don't know how the 
licensing works in that case.

---Dan

**Comment by CelticMinstrel** [comment]
It basically means we would need to trace back to the original artist and get their permission too, which is pretty much the same for any other images that we want to relicense.

**Comment by beetlenaut** [comment]
Thank you very much for your help! I'm working on this now, so let's not 
delete the PR.

---Dan

**Comment by beetlenaut** [comment]
I didn't get what you meant at first, but it should be fixed now.

---Dan

**Comment by CelticMinstrel** [comment]
I guess references to the removed images don't need to be updated? Just making sure here.

**Comment by CelticMinstrel** [comment]
I guess references to the removed images don't need to be updated? Just making sure here. (Also, I think chest-open can also be removed now, right?)

**Comment by beetlenaut** [comment]
The references are fine because I was using folders with the same names as in 
core. chest-open.png isn't in this branch, so it will cause an error if I 
remove it. I thought I should probably wait until after this is merged.

---Dan

**Comment by CelticMinstrel** [comment]
I'd prefer to rebase this and erase the removed images from history altogether, just for the record. But yeah, if it's still needed on this branch I guess it makes sense to leave it in for the time being.

**Comment by CelticMinstrel** [comment]
By the way, this has nothing to do with the PR or anything, but do you know how hard it would be to get a version of the ship facing an opposite direction? (The directions not obtainable via simple horizontal flip.)

File: data/campaigns/Secrets_of_the_Ancients/scenarios/02_Dark_Business.cfg [hunk 1]
```diff
@@ -227,12 +227,13 @@ I decided to hide in the cemetery. That way I could try my experiment to animate
...
-				description= _ "This entire scenario takes place at night."
+				description= _ "Right-click on a grave adjacent to your leader to get a unit."
+			[/note]
+			[note]
+				description= _ "Each grave can only be used once."
 			[/note]
 			[note]
-				red,blue,green=0,255,255
-				description= _ "Right-click on nearby gravestones to get more units."
+				description= _ "This entire scenario takes place at night."
```
Comment (GregoryLundberg) [inline data/campaigns/Secrets_of_the_Ancients/scenarios/02_Dark_Business.cfg]: This scenario takes place entirely at night.

File: data/campaigns/Secrets_of_the_Ancients/scenarios/02_Dark_Business.cfg [hunk 2]
```diff
@@ -227,12 +227,13 @@ I decided to hide in the cemetery. That way I could try my experiment to animate
 			[/gold_carryover]
 			
 			[note]
-				red,blue,green=0,255,255
-				description= _ "This entire scenario takes place at night."
+				description= _ "Right-click on a grave adjacent to your leader to get a unit."
+			[/note]
+			[note]
+				description= _ "Each grave can only be used once."
```
Comment (Vultraz) [inline data/campaigns/Secrets_of_the_Ancients/scenarios/02_Dark_Business.cfg]: There isn't really a difference, and the original is more colloquial.

File: data/campaigns/Secrets_of_the_Ancients/scenarios/02_Dark_Business.cfg (hunk 1 above)
Comment (CelticMinstrel) [inline data/campaigns/Secrets_of_the_Ancients/scenarios/02_Dark_Business.cfg]: That does sound better, yes (though the original _is_ grammatical).

**Comment by beetlenaut** [comment]
I decided to change both lines.

**Comment by beetlenaut** [comment]
Oh, right. You did mention that before. It sounds nice, but I'm new at this, 
and I doubt my ability to make it happen correctly. It would be better if 
someone else handles it.

I still have two changes I need to make on one scenario, but then I will be 
done with everything I planed to fix and the branch should be ready. I should 
be able to finish by tomorrow.

---Dan

**Comment by beetlenaut** [comment]
After a flip, the shading/shadows don't match the normal BfW direction. That 
wouldn't take too much work to fix, but a couple hours anyway.

Rewriting the terrain masks would take about as long though. They're tricky.

I think I will plan to do that eventually though. If the ship faced the other 
way in Becalmed, it would match the bigmap better.

---Dan

**Comment by CelticMinstrel** [comment]
Just to clarify: The ship as it stands is currently moving in a southerly direction. I'd like a version moving in a northerly direction, hence why the simple horizontal flip is not sufficient. (A horizontal flip swaps west and east, but it's still southerly.)

**Comment by beetlenaut** [comment]
Ah, I see. I didn't read carefully enough. That would require redrawing about 
80% of it. I'm pretty slow, and it took me a couple weeks to do the current 
one, so I'm not likely to work on another.

I could put the layered gimp file somewhere so it would be easier for someone 
to pick out the bits that are reusable.

---Dan

**Comment by CelticMinstrel** [comment]
> That would require redrawing about 80% of it.

You might be correct, but I do kind of wonder if a "skew" type operation would be able to produce the desired effect. Haven't yet found time to try that (mainly because the campaign I want to use it for is still in its infancy).

To clarify what I mean by skew, it'd be something like this:

```
oo7      1oo
o48      24o
159  --> 357
26o      o68
3oo      oo9
```

Where "o" are irrelevant (probably transparent) pixels and 1-9 represent the important pixels. I'm not going to ask you to do this for me (I'm sure I could manage it myself if I wanted), but if you have any opinions on whether that might work, I wouldn't mind hearing it.

**Comment by beetlenaut** [comment]
The "shear" command in gimp does exactly what you are showing here (and you 
can probably find that transformation in any other drawing program as well). 
It only takes a quick test to show you that skewing it just distorts it, and 
way worse than you probably thought it would.

I don't think there is any shortcut here. The parts of the ship that we would 
need to see if it was sailing away just aren't in this drawing.

---Dan

**Comment by beetlenaut** [comment]
As far as I know, the only thing left to do before this branch is merged (or rebased?) is to run wmlindent on it. However, I'm holding off for the moment in case someone notices another problem. (I hate working with space-indented files.)

Nobody has mentioned the Travis-CI failures, so am I right in assuming they aren't that important? They don't seem to correspond to anything. For example, I did nothing except add a line to .gitignore, and every unit test failed for that commit.

**Comment by gfgtdf** [comment]
There are some things:
1) Afaik the {FOREACH} maor is deprecated, consider replacing it with [for] or [foreach]
2) Some files are missing trailing empty line, maybe wmlindent will automatically fix this, not sure.
3) When i looked at this code i found some #ifdef s that could be simplyfied by using ON_DIFFICULTY or QUANTITY macro.


> Nobody has mentioned the Travis-CI failures, so am I right in assuming they aren't that important

Travis doesn't test the campaigns content so we assume its unrelated to the content of this pr and a bug in current master.

**Comment by CelticMinstrel** [comment]
Merge in process.

**Comment by CelticMinstrel** [comment]
Is there a reason to use logo.png rather than the identical-except-for-size wesnoth-icon.png?

**Comment by CelticMinstrel** [comment]
I've squashed things together (roughly by author, meaning two commits), run wmlindent and wmllint, and fixed up a few other deprecated things. You can view the result in the [Secrets_of_the_Ancients](https://github.com/wesnoth/wesnoth/tree/Secrets_of_the_Ancients) branch. It'll probably be merged into master tomorrow or so, but I'd like to hear your reasoning on the existence of logo.png first. If that can be replaced with an image already in mainline, so much the better.

**Comment by beetlenaut** [comment]
Oops. You can ignore that last one I guess. I didn't notice that a merge was in progress.

**Comment by beetlenaut** [comment]
The size was the reason.

---Dan

**Comment by CelticMinstrel** [comment]
So you needed it bigger for some reason?

**Comment by CelticMinstrel** [comment]
(It seems like either wmlindent or wmllint adds the newline at end of file - the branch already has that fixed, in any case.)

**Comment by sevu** [comment]
I had a look at it too, some questions.
- what about the slightly different mainline zombie portrait instead of the current one in scenario 2?
- recently merfolk castles and encampments were added. Could they be of use here too?  There is a map naming the merfolk, but I didn't look at it since I don't want to spoiler me to much.  
- the dwarvish miner in _sof_ did get an update a year ago - there is an new image for him having his backpack loaded. Is there some use for that here too? (the attack/defense animations are already added)
- _sotbe_ has two more orcish shamans. Any use here?
- Is it intended that Sailors do not advance to Sea Captains?
- Are there any problems with the Ruffian being the male version of the Delinquent? (but still having different advancement paths)
- And last but not least: The bats have to have layer=0 in _some_ scenario. Do the unit types get transformed for that scenarios? (like, with [transform_unit] or similar) Could we change the layer with an event or an object instead? This would have the advantage that we're not having the choice between having two units in the wiki or not showing their advancement path in the wiki. (the problem with hide_help)

**Comment by beetlenaut** [comment]
No, I didn't "need," it bigger, I just thought it looked better that way.

---Dan

**Comment by beetlenaut** [comment]
On Wednesday, December 14, 2016 8:44:51 PM Shiki wrote:
> - what about the slightly different mainline zombie portrait instead of the
> current one in scenario 2?
The mainline zombie clearly died in a battle (there's an arrow shaft through 
his arm and some large wounds), while the one in SotA is not intended to be a 
soldier.

> - recently merfolk castles and encampments were
> added. Could they be of use here too?
I forgot about that! I meant to use those. I was going to use a merfolk castle 
on scenario 10, and an encampment on scenario 21.

> - the > dwarvish miner in _sof_ did get an update a year ago
> - there is an new  image for him having his backpack loaded.
I didn't know about that. There is a miner with a line of dialog, so he could 
use that image. Maybe the unit should be added to core first though since it 
would be in two campaigns.

> - _sotbe_ has two more orcish shamans. Any use here?
I don't think so.

> - Is it intended that Sailors do not advance to Sea Captains?
Yes. There can only be one captain!

> - Are there any problems with the Ruffian being the male version of the
> Delinquent? (but still having different advancement paths)
I can't see any problem with it. Did you have some problem in mind? (The 
delinquent is actually a bit different than the ruffian too--it has a weaker 
melee attack but also a ranged attack.)

> - And last but not least: The bats have to have layer=0 in _some_ scenario. 
> Do the unit types get transformed for that scenarios?
Yes.

> Could we change the layer with an event or an object instead? 
As far as I know, there is no way to modify the standing animation (and thus 
the layer) from inside a scenario.

---Dan

**Pull request closed without merging**